# Position Matters: Statistical Profiles Across the PLL - Data Pre-Processing

This is all the data cleaning and pre-processing done for the EDA of Premier Lacrosse League Data.

## Data Pre-processing
Data sourced from *Premier Lacrosse League* website's player stats for 2025 Regular Season (https://stats.premierlacrosseleague.com/player-table).

A significant amount of time is taken to pre-process the data. It is an important aspect of any data analysis.

### Import necessary libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

### Import datasets

In [2]:
raw_data = pd.read_csv("pll-player-stats.csv")
raw_data

,First Name,Last Name,Position,Jersey,Team,Games Played,Points,Scoring Points,Total Goals,1Pt Goals,...,2Pt Goals Against,Penalties,Penalties (in minutes),Power Play Goals,Power Play Shots,Power Play Goals Against,Short Handed Goals,Short Handed Shots,Unassisted Goals,Assisted Goals
0,Connor,Shellenberger,A,1.0,Atlas,10,46,23,23,23,...,0,2,2.0,1,1,0,0,0,12,11
1,Michael,Sowers,A,22.0,Waterdogs,10,41,18,18,18,...,0,0,0.0,1,1,0,0,0,14,4
2,Jeff,Teat,A,7.0,Atlas,10,40,17,17,17,...,0,0,0.0,1,4,0,0,0,9,8
3,Pat,Kavanagh,A,51.0,Outlaws,10,37,21,21,21,...,0,7,5.0,0,5,0,0,0,9,12
4,Chris,Kavanagh,A,96.0,Redwoods,10,37,25,23,21,...,0,1,2.0,0,1,0,0,0,11,12
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
246,James,Shipley,M,19.0,Archers,0,0,0,0,0,...,0,0,0.0,0,0,0,0,0,0,0
247,John,Geppert,LSM,7.0,Cannons,0,0,0,0,0,...,0,0,0.0,0,0,0,0,0,0,0
248,Bubba,Fairman,SSDM,2.0,Cannons,0,0,0,0,0,...,0,0,0.0,0,0,0,0,0,0,0
249,TJ,Comizio,SSDM,33.0,Cannons,0,0,0,0,0,...,0,0,0.0,0,0,0,0,0,0,0


In [3]:
#copy contents of data into new variable to manipulate it
df = raw_data.copy()
df

,First Name,Last Name,Position,Jersey,Team,Games Played,Points,Scoring Points,Total Goals,1Pt Goals,...,2Pt Goals Against,Penalties,Penalties (in minutes),Power Play Goals,Power Play Shots,Power Play Goals Against,Short Handed Goals,Short Handed Shots,Unassisted Goals,Assisted Goals
0,Connor,Shellenberger,A,1.0,Atlas,10,46,23,23,23,...,0,2,2.0,1,1,0,0,0,12,11
1,Michael,Sowers,A,22.0,Waterdogs,10,41,18,18,18,...,0,0,0.0,1,1,0,0,0,14,4
2,Jeff,Teat,A,7.0,Atlas,10,40,17,17,17,...,0,0,0.0,1,4,0,0,0,9,8
3,Pat,Kavanagh,A,51.0,Outlaws,10,37,21,21,21,...,0,7,5.0,0,5,0,0,0,9,12
4,Chris,Kavanagh,A,96.0,Redwoods,10,37,25,23,21,...,0,1,2.0,0,1,0,0,0,11,12
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
246,James,Shipley,M,19.0,Archers,0,0,0,0,0,...,0,0,0.0,0,0,0,0,0,0,0
247,John,Geppert,LSM,7.0,Cannons,0,0,0,0,0,...,0,0,0.0,0,0,0,0,0,0,0
248,Bubba,Fairman,SSDM,2.0,Cannons,0,0,0,0,0,...,0,0,0.0,0,0,0,0,0,0,0
249,TJ,Comizio,SSDM,33.0,Cannons,0,0,0,0,0,...,0,0,0.0,0,0,0,0,0,0,0


### Check for null data

In [4]:
#check data types of the imported file to ensure accurate and complete import
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 251 entries, 0 to 250
Data columns (total 45 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   First Name                251 non-null    object 
 1   Last Name                 251 non-null    object 
 2   Position                  251 non-null    object 
 3   Jersey                    246 non-null    float64
 4   Team                      251 non-null    object 
 5   Games Played              251 non-null    int64  
 6   Points                    251 non-null    int64  
 7   Scoring Points            251 non-null    int64  
 8   Total Goals               251 non-null    int64  
 9   1Pt Goals                 251 non-null    int64  
 10  2Pt Goals                 251 non-null    int64  
 11  Assists                   251 non-null    int64  
 12  Shots                     251 non-null    int64  
 13  Shooting %                251 non-null    float64
 14  Shots On G

There are 251 rows of data for each of 45 columns of data. This means there are 251 players in the Premier Lacrosse League in 2025. There are only 5 missing values in the `Jersey` field, so the data is mostly complete.

In [5]:
#investigate the missing data values
df[df["Jersey"].isna()]

,First Name,Last Name,Position,Jersey,Team,Games Played,Points,Scoring Points,Total Goals,1Pt Goals,...,2Pt Goals Against,Penalties,Penalties (in minutes),Power Play Goals,Power Play Shots,Power Play Goals Against,Short Handed Goals,Short Handed Shots,Unassisted Goals,Assisted Goals
200,Devon,McLane,A,NaN,Archers,0,0,0,0,0,...,0,0,0.0,0,0,0,0,0,0,0
222,Scott,Cole,A,NaN,Whipsnakes,0,0,0,0,0,...,0,0,0.0,0,0,0,0,0,0,0
233,Griffin,Schutz,M,NaN,Redwoods,0,0,0,0,0,...,0,0,0.0,0,0,0,0,0,0,0
234,Will,Lynch,FO,NaN,Archers,0,0,0,0,0,...,0,0,0.0,0,0,0,0,0,0,0
239,Pace,Billings,D,NaN,Outlaws,0,0,0,0,0,...,0,0,0.0,0,0,0,0,0,0,0


In [6]:
df.loc[[200,222,233,234,239],["First Name","Last Name", "Position", "Games Played"]]

,First Name,Last Name,Position,Games Played
200,Devon,McLane,A,0
222,Scott,Cole,A,0
233,Griffin,Schutz,M,0
234,Will,Lynch,FO,0
239,Pace,Billings,D,0


Note that these players did not play in any games. This means they were on the roster but never played. They do not provide any insight into player statistics because they have no statistics. For the purposes of this analysis, we will *exclude* them.

In [7]:
df_clean = df[df["Games Played"]>0].copy()

### Check positional statistics

Now, investigate the statistics that are positional. We need to determine whether Face-off and Goalie statistics are genuinely zero for non-specialists, or if the zeroes stand in for "Not Applicable."

In [8]:
df_clean.groupby("Position")[["Faceoffs","Shots", "Saves"]].mean()

,Faceoffs,Shots,Saves
Position,,,
A,0.000000,42.828571,0.000000
D,0.033333,2.500000,0.000000
FO,193.800000,9.800000,0.000000
G,0.000000,1.000000,62.705882
LSM,0.000000,4.600000,0.000000
M,0.980769,27.288462,0.000000
SSDM,4.000000,4.451613,0.000000


Observe that these statistics are legitimate. Attackmen (`A`) rarely play in faceoffs or as a goalie, so their statistics as a zero are realistic. Defensemen (`D`) sometimes play faceoffs, which explains the nonzero statistic there. Goalies (`G`) can sometimes shoot, but their role primarily consists of saving shots. 

### Check for duplicates and team name datatypes

In [9]:
#Check if there are duplicated rows
print("Duplicated rows:", df_clean.duplicated().sum())

#Check if team names are consistent
print("Unique Team Names:", df_clean["Team"].nunique())

Duplicated rows: 0
Unique Team Names: 9


There are no duplicated rows, which is good. No action needed.

There are only eight teams in the PLL, so we investigate why there are nine values present.

In [10]:
df_clean["Team"].unique()

array(['Atlas', 'Waterdogs', 'Outlaws', 'Redwoods', 'Cannons',
       'Whipsnakes', 'Chaos', 'Archers', 'Player Pool'], dtype=object)

There is an extra `Team` called `Player Pool`. In the PLL, the Player Pool consists of all active roster players, free agents, and recent college draftees. These are players who do not have a contract with a specific team at the time the statistics were collected. We investigate to see if there are players who have actually played games during the season in it.

In [11]:
df_clean[df_clean["Team"]=="Player Pool"][["First Name", "Last Name", "Team", "Games Played"]]

,First Name,Last Name,Team,Games Played
124,CJ,Costabile,Player Pool,1
130,Gannon,Matthews,Player Pool,1
168,Dylan,Renner,Player Pool,1


The three players in the Player Pool category have played in a game each. Hence, we will include them in the Player Analysis. However, during the Team Analysis, the `Player Pool` team will be excluded because it is not a full team roster.


### Sanity check percentage columns

In [12]:
print(df_clean.columns.tolist())

['First Name', 'Last Name', 'Position', 'Jersey', 'Team', 'Games Played', 'Points', 'Scoring Points', 'Total Goals', '1Pt Goals', '2Pt Goals', 'Assists', 'Shots', 'Shooting %', 'Shots On Goal', 'Shots On Goal %', '2Pt Shots', '2Pt Shots %', '2Pt Shots On Goal', '2Pt Shots On Goal %', 'Turnovers', 'Caused Turnovers', 'Groundballs', 'Touches', 'Passes', 'Faceoff %', 'Faceoffs Won', 'Faceoffs Lost', 'Faceoffs', 'Scores Against Average', 'Saves', 'Save %', 'Clean Saves', 'Clean Save %', 'Scores Against', '2Pt Goals Against', 'Penalties', 'Penalties (in minutes)', 'Power Play Goals', 'Power Play Shots', 'Power Play Goals Against', 'Short Handed Goals', 'Short Handed Shots', 'Unassisted Goals', 'Assisted Goals']


In [13]:
#name the columns that are supposed to be percentage values
pct_columns = ["Shooting %", "Shots On Goal %", "2Pt Shots %", "2Pt Shots On Goal %", "Faceoff %", "Save %", "Clean Save %"]
percent_columns = df_clean[pct_columns]

for col in percent_columns:
    print(col, df_clean[col].min(), df_clean[col].max())

Shooting % 0.0 1.0
Shots On Goal % 0.0 1.0
2Pt Shots % 0.0 1.0
2Pt Shots On Goal % 0.0 1.0
Faceoff % 0.0 1.0
Save % 0.0 0.625
Clean Save % 0.0 0.6055045871559633


These values stay between 0 and 1, which is the requirement for percentages. No action needed.

### Last few sanity checks

In [14]:
#Sanity check total goals is 1 point + 2 point goals

print("Total Goals Check:", (df_clean["Total Goals"] != (df_clean["1Pt Goals"] + df_clean["2Pt Goals"])).sum())

#Sanity check points is number of 1 pt goals + 2*number of 2 pt goals + assists
print("Points Check:", (df_clean["Points"] != (df_clean["1Pt Goals"] + 2*df_clean["2Pt Goals"] + df_clean["Assists"])).sum())

Total Goals Check: 0
Points Check: 0


We ensure the calculations for total goals and total points are the expected functions for interpretation. They are, so no action is needed.

### Export new cleaned dataset to csv
Now the data is cleaned and prepared for our analysis, we can export the modified file to use.

In [15]:
df_clean.to_csv("pll-player-stats-cleaned.csv", index = False)